icu4j/main/core/src/main/java/com/ibm/icu/util/ULocale.java

```diff
@@ -581,35 +581,48 @@ public Locale toLocale() {
      * @stable ICU 2.8
      */
     public static ULocale getDefault() {
+        // Only synchronize if we must update the default locale.
+        if (Objects.equals(defaultLocale, Locale.getDefault())) {
+            return getNonNullULocale(defaultULocale);
```

Comment:
I think that if both the defaultLocale and the Locale have changed and the return hasn't been executed then it's all good, that's because when we modify both variables we do them in this order:

defaultULocale = nextULocale;
defaultLocale = currentDefault;

So if we enter the if statement the defaultULocale will already have the correct value in this case.

With that said, if we enter the if statement and the Locale changes (which is a legit scenario), the existing code wouldn't catch that anyways given that we have already cached its value.

What do you think?

Reply:
I think the compiler is free to reorder unsynchronized code for optimizations, unless there is a sequential dependency, like one line using a value computed on a preceding line. So I am not convinced that the current version is safe.

Reply:
Hmm, I've been deep diving into the implementation details of Java concurrency and I've found the following:

https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/util/concurrent/package-summary.html

"Each action in a thread happens-before every action in that thread that comes later in the program's order."

"A write to a volatile field happens-before every subsequent read of that same field. Writes and reads of volatile fields have similar memory consistency effects as entering and exiting monitors, but do not entail mutual exclusion locking."

Unless I'm misunderstanding the meaning of action, the assumption that the order is fixed should be true, is this not the case?

Anyways, I'll think more about this and modify the code tomorrow.

Reply:
I thought about it and came up with this fix:

Cache the defaultULocale, otherwise it might become null and we would throw a NPE later on when we try to get the locale out of it.

I put some comments explaining that in the code.

Let me know if this approach works for you.

Reply:
I think there was some slight confusion here, the current version of the code is already doing exactly what you describe in the code snipper at t he end of the comment, I got rid of the defaultLocale since the last iteration.